pageserver/src/http/routes.rs

```diff
@@ -31,24 +34,45 @@ struct State {
     auth: Option<Arc<JwtAuth>>,
     remote_index: RemoteIndex,
     allowlist_routes: Vec<Uri>,
+    remote_storage: Option<GenericRemoteStorage>,
+}
+
+enum GenericRemoteStorage {
+    Local(LocalFs),
+    S3(S3Bucket),
 }
 
 impl State {
     fn new(
         conf: &'static PageServerConf,
         auth: Option<Arc<JwtAuth>>,
         remote_index: RemoteIndex,
-    ) -> Self {
+    ) -> anyhow::Result<Self> {
         let allowlist_routes = ["/v1/status", "/v1/doc", "/swagger.yml"]
             .iter()
             .map(|v| v.parse().unwrap())
             .collect::<Vec<_>>();
-        Self {
+        let remote_storage = conf
```

Comment:
Yes, the generics are the culprit: we cannot actually implement `RemoteStorage` for `GenericTimeline` enum, because of the `StoragePath` associated item being different: so, we'll have to match against every enum variant across all code where it uses it. Might be doable, but will bloat this PR quite significantly.

Throwing away `GenericTimeline` in `routes.rs` is not viable too, since it will poison `State` and the rest of the methods with generics.

Reply:
I do not have a strong opinion here. State with generics will be quite lengthy but looks ok to me. The current approach benefits in readability via minor duplication, maybe worth a comment though

Reply:
Added.